framework/src/utils/BootstrapStatistics.C

```diff
@@ -93,59 +97,104 @@ BootstrapCalculator::shuffle(const std::vector<Real> & data,
+    std::vector<std::size_t> offsets(n_processors());
+    {
+      std::vector<std::size_t> local_sizes;
+      _communicator.allgather(n_local, local_sizes);
+      for (std::size_t i = 1; i < local_sizes.size(); ++i)
+        offsets[i] = offsets[i-1] + local_sizes[i];
+    }
+
+
+    {
+      SerializerGuard sg(_communicator, false);
+      std::cerr << "data " << processor_id() << " = {";
+      for (auto val : data)
+        std::cerr << val << " ";
+      std::cerr << "}" << std::endl;
+
+      std::cerr << "offset " << processor_id() << " = {";
+      for (auto val : offsets)
+        std::cerr << val << " ";
+      std::cerr << "}" << std::endl;
+    }
 
-    /*
-    std::cerr << processor_id() << " offset = {";
-    for (auto val : offsets)
-      std::cerr << val << " ";
-    std::cerr << "}" << std::endl;
-    */
 
     // Advance the random number generator
     for (std::size_t i = 0; i < offsets[processor_id()]; ++i)
-      generator.randl(0, 0, global_n);
+      generator.randl(0, 0, n_global);
 
     // Compute the needs for this processor
-    std::vector<std::vector<std::size_t>> needs(n_processors());
-    for (std::size_t j = 0; j < n; ++j)
+    std::unordered_map<processor_id_type, std::vector<std::size_t>> needs;
+    for (std::size_t i = 0; i < n_local; ++i)
     {
-      auto index = generator.randl(0, 0, global_n);
-      auto rank = std::lower_bound(offsets.begin(), offsets.end(), index);
-      auto rk = std::distance(offsets.begin(), rank);
+      auto index = generator.randl(0, 0, n_global);
+
+      // iterator to first element in range that is GREATER THAN index
+      auto rank = std::upper_bound(offsets.begin(), offsets.end(), index);
+      auto rk = std::distance(offsets.begin(), rank) - 1;
 
-      needs[rk].push_back(index);
+      /*
+      std::size_t rk = 0;
+      for (std::size_t r = 0; r < offsets.size(); ++r)
+        if (index > offsets[r])
+        {
+          rk = r;
+          break;
+        }
+      */
+
+      needs[rk].push_back(index - offsets[rk]);// - offsets[processor_id()]);
 
       //indices[j] = index;//generator.randl(0, 0, global_n);
     }
 
+    {
+      SerializerGuard sg(_communicator, false);
+      std::cerr << "n_local = " << n_local << std::endl;
+      std::cerr << "RANK: " << processor_id() << std::endl;
+      for (auto iter : needs)
+      {
+        std::cerr << " needs[" << iter.first << "] = {";
+        for (auto val : iter.second)
+          std::cerr << val << " ";
+        std::cerr << "}" << " " << offsets[iter.first] << std::endl;
+      }
+    }
+
+    //
+    std::unordered_map<processor_id_type, std::vector<Real>> returns;
+    auto return_functor = [&data, &returns](processor_id_type  pid,
+                                            const std::vector<std::size_t> & indices) {
+                            for (auto idx : indices)
+                              returns[pid].push_back(data[idx]);
```

Comment:
Because this is a map you should `.at(pid)` outside of the for loop and access the entry as such instead doing it for every entry

Reply:
...I know it's a WIP but I was just curious

Reply:
I highly discourage the use of `at()` because it throws an exception that doesn't gracefully exit (e.g. call MPI_Abort) when running in parallel. It too bad because semantically, that is indeed the right method to call in many cases. Instead you have to fall back to `.find()` or `.count()` depending on the context. In any case, you definitely want to avoid "accidental insertion", which was a favorite error in MOOSE until I rooted most if not all of them out.

Reply:
Turns out `at()` was actually the wrong suggestion here anyway - we want to insert if it doesn't exist. However: definitely good to know.